runtime/src/iree/base/internal/cpu.c

```diff
@@ -10,9 +10,144 @@
 #include "iree/base/internal/cpu.h"
 
 #include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
 
 //===----------------------------------------------------------------------===//
-// iree_cpu_*
+// Platform-specific processor data queries
+//===----------------------------------------------------------------------===//
+
+#if defined(IREE_PLATFORM_ANDROID) || defined(IREE_PLATFORM_LINUX)
+
+// NOTE: not all kernel versions have all of the cap bits we need defined so as
+// a practice we always define the feature bits we need locally.
+#include <sys/auxv.h>
+
+// OR's |field_bit| into |field_value| if |hwcap_bit| is set in |hwcap_value|.
+#define IREE_SET_IF_HWCAP(hwcap_value, hwcap_bit, field_value, field_bit)   \
+  (field_value) = (field_value) | iree_all_bits_set(hwcap_value, hwcap_bit) \
+                      ? field_bit                                           \
+                      : 0
+
+#if defined(IREE_ARCH_ARM_64)
+// https://docs.kernel.org/arm64/elf_hwcaps.html
+#define IREE_HWCAP_ASIMD (1 << 1)
+#define IREE_HWCAP2_I8MM (1 << 13)
+static void iree_cpu_query_data_arch_hwcaps(uint32_t hwcap, uint32_t hwcap2,
+                                            uint64_t* out_fields) {
+  IREE_SET_IF_HWCAP(hwcap, IREE_HWCAP_ASIMD, out_fields[0],
+                    IREE_CPU_DATA_FIELD_0_AARCH64_HAVE_ASIMDDP);
+  IREE_SET_IF_HWCAP(hwcap2, IREE_HWCAP2_I8MM, out_fields[0],
```

Comment:
If we rely on explicit `HWCAP` bits to detect such recent CPU features then we risk running again into the trouble we were in with the Pixel4 launch where it was only weeks ahead of the public device launch that an OS image update brought a Linux kernel backport exposing the `HWCAP` feature bit for the `DP` extension.

My lesson learned from that episode is that it's better to rely only on `HWCAP_CPUID` and use trapped access to CPU id registers (https://docs.kernel.org/arm64/cpu-feature-registers.html).  This still relies a sufficiently recent kernel version to have `HWCAP_CPUID`, but at least from there on, the kernel version requirement is frozen, so for cutting-edge features such as `i8mm` that is a win. As the Pixel4 experience showed, Android kernel versions on brand new Pixel devices can happily trail upstream by years.

https://github.com/iree-org/iree/blob/main/runtime/src/iree/tooling/cpu_features.c shows an example of using the `HWCAP_CPUID` way.

It's totally cool if you're OK with the idea but still want to merge this as-is and defer to me to do that as a follow-up - just confirm if we're aligned on the strategy :-)

Reply:
I can also live with the explicit `HWCAP` feature flags for now as long as my moto device has a recent enough kernel (i need to check) and as long as we revisit before someone tries to ship something to real users (some of whom will be running a kernel not exposing useful features in `HWCAP` bits)

Reply:
as noted above I'm OK to refactor/merge/dedupe that as a followup after the present PR is merged, if we agree on the approach.

Reply:
I'm going to leave it for the moment because I don't have a system to test on, but I did add a TODO noting that we should make use of it!

Reply:
SG - i'll take care of it next week.